pkg/gencred/gencred.go

```diff
@@ -28,23 +28,122 @@ import (
 	"time"
 
 	certificates "k8s.io/api/certificates/v1beta1"
-	v1 "k8s.io/api/core/v1"
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/util/wait"
 	"k8s.io/client-go/kubernetes"
+	_ "k8s.io/client-go/plugin/pkg/client/auth"
 	"k8s.io/client-go/util/cert"
 	"k8s.io/client-go/util/keyutil"
 )
 
 const (
+	// clusterRoleName is the name for the superuser ClusterRole.
+	clusterRoleName = "cluster-admin"
+	// clusterRoleBindingName is the name for the cluster administrator ClusterRoleBinding.
+	clusterRoleBindingName = "serviceaccount-cluster-admin-crb"
+	// serviceAccountName is the name for the cluster administrator ServiceAccount.
+	serviceAccountName = "serviceaccount-cluster-admin"
 	// systemPrivilegedGroup is a superuser by default (i.e. bound to the cluster-admin ClusterRole).
 	systemPrivilegedGroup = "system:masters"
-	// certWaitInterval certificate request poll interval.
-	certWaitInterval = time.Second
-	// certWaitTimeout certificate request poll timeout.
-	certWaitTimeout = 20 * time.Second
+	// waitInterval request poll interval.
+	waitInterval = time.Second
+	// waitTimeout request poll timeout.
+	waitTimeout = 20 * time.Second
 )
 
+// getOrCreateSA gets existing or creates new service account (SA).
+func getOrCreateSA(clientset kubernetes.Interface) ([]byte, []byte, error) {
+	client := clientset.CoreV1().ServiceAccounts(corev1.NamespaceDefault)
+
+	// Get ServiceAccount if exists.
+	if saObj, err := client.Get(serviceAccountName, metav1.GetOptions{}); err == nil {
+		return getSASecrets(clientset, saObj)
+	}
+
+	// Generate a Kubernetes ServiceAccount object.
+	saObj := &corev1.ServiceAccount{
+		TypeMeta: metav1.TypeMeta{Kind: "ServiceAccount"},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: serviceAccountName,
+		},
+	}
+
+	// Create ServiceAccount.
+	saObj, err := client.Create(saObj)
+	if err != nil {
+		return nil, nil, fmt.Errorf("create SA: %v", err)
+	}
+
+	// Get/Create ClusterRoleBinding.
+	err = getOrCreateCRB(clientset)
+	if err != nil {
+		return nil, nil, fmt.Errorf("get or create CRB: %v", err)
+	}
+
+	// Get ServiceAccount.
+	saObj, err = client.Get(serviceAccountName, metav1.GetOptions{})
+	if err != nil {
+		return nil, nil, fmt.Errorf("get SA: %v", err)
+	}
+
+	return getSASecrets(clientset, saObj)
+}
+
+// getOrCreateCRB gets existing or creates new cluster role binding (CRB).
+func getOrCreateCRB(clientset kubernetes.Interface) error {
+	client := clientset.RbacV1().ClusterRoleBindings()
+
+	// Get ClusterRoleBinding if exists.
+	if _, err := client.Get(clusterRoleBindingName, metav1.GetOptions{}); err == nil {
+		return nil
+	}
+
+	// Generate a Kubernetes ClusterRoleBinding object.
+	crbObj := &rbacv1.ClusterRoleBinding{
+		TypeMeta:   metav1.TypeMeta{Kind: "ClusterRoleBinding"},
```

Comment:
Nit: You do not need to set the `TypeMeta`, this is done automatically during serialization

Reply:
That is good to know. I have cleanup up the superfluous use of `TypeMeta`.